engines/python/setup/djl_python/huggingface.py

```diff
@@ -496,41 +496,96 @@ def register_adapter(inputs: Input):
     """
     Registers lora adapter with the model.
     """
+    if not is_rolling_batch_enabled(_service.hf_configs.rolling_batch):
+        raise NotImplementedError(
+            "LoRA adapter API is only supported for rolling batch.")
+
     adapter_name = inputs.get_property("name")
     adapter_path = inputs.get_property("src")
+    adapter_pin = inputs.get_property("pin") == "true"
     if not os.path.exists(adapter_path):
         raise ValueError(
             f"Only local LoRA models are supported. {adapter_path} is not a valid path"
         )
     logging.info(f"Registering adapter {adapter_name} from {adapter_path}")
     _service.adapter_registry[adapter_name] = inputs
+
+    result = _service.rolling_batch.add_lora(adapter_name, adapter_path)
+    if adapter_pin:
+        try:
+            if not result:
+                return Output().error(
+                    "error",
+                    message=f"Failed to register LoRA adapter {adapter_name}")
+            result = _service.rolling_batch.pin_lora(adapter_name)
+            if not result:
+                return Output().error(
+                    "error",
+                    message=f"Failed to register LoRA adapter {adapter_name}")
+        except (ValueError, RuntimeError) as e:
+            if any(msg in str(e)
+                   for msg in ("No free lora slots",
+                               "greater than the number of GPU LoRA slots")):
+                raise MemoryError(str(e))
+            return Output().error(
+                f"Failed to register LoRA adapter {adapter_name}",
+                message=str(e))
+
+    logging.info(
+        f"Registered adapter {adapter_name} from {adapter_path} successfully")
+    return Output(message=f"Adapter {adapter_name} registered")
+
+
+def update_adapter(inputs: Input):
+    """
+    Updates lora adapter with the model.
+    """
     if not is_rolling_batch_enabled(_service.hf_configs.rolling_batch):
-        if isinstance(_service.model, PeftModel):
-            _service.model.load_adapter(adapter_path, adapter_name)
-        else:
-            _service.model = PeftModel.from_pretrained(_service.model,
-                                                       adapter_path,
-                                                       adapter_name)
+        raise NotImplementedError(
+            "LoRA adapter API is only supported for rolling batch.")
 
-        if isinstance(_service.hf_pipeline, Pipeline):
-            _service.hf_pipeline.model = _service.model
+    adapter_name = inputs.get_property("name")
+    adapter_path = inputs.get_property("src")
+    adapter_pin = inputs.get_property("pin") == "true"
+    if adapter_name not in _service.adapter_registry:
+        raise ValueError(f"Adapter {adapter_name} not registered.")
+    _service.adapter_registry[adapter_name] = inputs
```

Comment:
If the src changes, is it sufficient to just update the adapter registry? Or do we need to unregister the previous adapter and re-register this one?

Reply:
let's add a validation that src is the same from what we have already, and what is now provided in the inputs

Reply:
raise NotImplementedError